Add "create a site" deeplink handling

planarvoid opened the pull request:
Fixes #14371 

This solution introduces handling for the `public-api.wordpress.com/mbar/...redirect_to=wordpress.com/wp-login.php?...redirect_to="wordpress.com/start"` links from emails and opens a site creation flow when these links are clicked.
There is now a view model which handles all the `mbar` links and all the links that open the editor (just to have them in one place). 

To test - Site creation:
- Log in in the app with a WP.com account
- Click on the "Create your site" in an email (let me know if you need the link from the email)
- Notice this line of code in the LogCat - "DeepLink tracking URI successfully requested"
- Notice the "Site creation" flow is started
- Finish the flow and create a site
- Notice you end up on the My Site screen with the site is selected 

To test 2 - Site creation not logged in
- Do not log in to the app with a WP.com account
- Click on the "Create your site" in an email (let me know if you need the link from the email)
- Notice this line of code in the LogCat - "DeepLink tracking URI successfully requested"
- Notice the "WP.com signup" flow is started
- Log in/signup
- You end up logged in on the My Site screen

To test 3 - edit an existing post
- Log in to the app
- Click on a link like `https://wordpress.com/post/site_url/post_id` that exists for a site on an account you're logged in with 
- Notice this line of code in the LogCat - "DeepLink tracking URI successfully requested"
- Notice the post is opened in the editor

To test 4 - edit a non-existing post
- Log in to the app
- Click on a link like `https://wordpress.com/post/site_url/post_id` that doesn't exists for a site on an account you're logged in with 
- Notice this line of code in the LogCat - "DeepLink tracking URI successfully requested"
- Notice the editor is opened for site with site_url

To test 5 - edit a non-existing post on a non-logged in site
- Log in to the app
- Click on a link like `https://wordpress.com/post/site_url/post_id` that doesn't exists for a site that you're not logged in
- Notice this line of code in the LogCat - "DeepLink tracking URI successfully requested"
- Notice the editor is opened for your current site

## Regression Notes
1. Potential unintended areas of impact
- All the `mbar` links are impacted but they should all be handled as before

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Unit tested the classes that could be tested

3. What automated tests I added (or what prevented me from doing so)
- unit tests in this PR

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

planarvoid committed:
First version of "Create a site" deeplink handling

planarvoid committed:
Add more fields to UriWrapper.kt

planarvoid committed:
Add EmailUriHandler.kt and test

planarvoid committed:
Use EmailUriHandler in DeepLinkingIntentReceiverActivity

planarvoid committed:
Add DeeplinkWebViewActivity.kt

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/DeeplinkWebViewActivity.kt:
```diff
@@ -0,0 +1,52 @@
+import org.wordpress.android.R
+import org.wordpress.android.ui.accounts.LoginMagicLinkInterceptActivity
+
+class DeeplinkWebViewActivity : WebViewActivity() {
```
If we decide to handle `mbar` links instead of `bar` ones, as we have been discussing recently, then I think we should be able to get rid of this class, right?

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/DeepLinkingIntentReceiverActivity.java:
```diff
@@ -91,14 +96,29 @@ protected void onCreate(Bundle savedInstanceState) {
             } else {
-                // not handled
-                finish();
+                UriWrapper uriWrapper = new UriWrapper(uri);
+                // https://public-api.wordpress.com/bar/
+                if (mEmailUriHandler.shouldHandleUri(uriWrapper)) {
+                    mEmailUriHandler.handleUri(uriWrapper);
+                } else {
+                    // not handled
+                    finish();
+                }
```
Again, if we decide to handle `mbar` links, I believe we should be able to merge this logic with the existing logic in `handleOpenEditorFromTrackingUrl` as they seem to do the similar things, right? Let me know what you think.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/DeeplinkNavigator.kt:
```diff
@@ -0,0 +1,32 @@
+                DeeplinkWebViewActivity.openUrl(activity, navigateAction.uri.toString())
+            }
+            is StartCreateSiteFlow -> {
+                if (navigateAction.isSignedToWpCom) {
+                    ActivityLauncher.newBlogForResult(activity)
+                } else {
+                    ActivityLauncher.addSelfHostedSiteForResult(activity)
+                }
```
Two thoughts here:

- If the user is not logged in to WP.com, I think it would probably be better for them to be redirected to the WP.com login flow instead of the add site flow.
- In addition to that, we should also check if the user is not logged in at all, in which case I think they should be redirected to the full login flow.

Let me know what you think!

planarvoid replied:
hmm, good point but I think we should only redirect to the WP.com flow because the link in the email only comes to WP.com users and I don't see a usecase where we'd like to redirect a user who got the "Create a site" email to the self-hosted flow. WDYT?

renanferrari reviewed (commented):
Thanks for submitting these changes!

I tested both the create site URL and the magic link URL and everything seems to be working as expected.

Code is also looking pretty good and it seems like a solid starting point for us! Left a few thoughts and questions below.

planarvoid committed:
Merge branch 'develop' into feature/14371-add-create-a-site-deeplink-handling

planarvoid committed:
Create DeepLinkingIntentReceiverViewModel.kt and handle email deeplinks there

planarvoid committed:
Add tests for deeplink view model

planarvoid committed:
Add server tracking and tests

planarvoid committed:
Simplify code and add tests

planarvoid committed:
Rename class

planarvoid committed:
Open site creation flow in a stack over the main activity

planarvoid committed:
Fix test

planarvoid commented:
thanks for the review @renanferrari , I've completely rewritten the whole solution to handle the `mbar` links, I've removed the WebView, tested the code, moved handling of all the `mbar` links and the editor links to the ViewModel, and overall improved the solution. I think this is ready now for another round.

planarvoid committed:
Remove activity from the manifest

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/DeepLinkNavigator.kt:
```diff
@@ -0,0 +1,47 @@
+
+    sealed class NavigateAction {
+        data class OpenInBrowser(val uri: UriWrapper) : NavigateAction()
+        data class OpenInEditor(val site: SiteModel, val postId: Int) : NavigateAction()
```
It's a nitpick, but maybe we could rename `OpenInEditor` to `OpenEditorForPost` for consistency with `ActivityLauncher::openEditorForPostInNewStack` and `OpenEditorForSite`?

planarvoid replied:
makes sense, changed in f32d99fcb4

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/DeepLinkingIntentReceiverViewModel.kt:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * URIs like `public-api.wordpress.com/mbar/...` come from emails and should be handled here
+     */
+    fun shouldHandleEmailUrl(uri: UriWrapper): Boolean {
```
These links currently come from emails, but in the future they might come from somewhere else. Wouldn't be better to keep the `shouldHandleTrackingUrl` name (as it was before)? Then we could also update the comment to indicate that this _usually_ comes from emails. Same thing for `handleEmailUrl` → `handleTrackingUrl`.

planarvoid replied:
tracking URL is a better name than email URL, changed in 0730a8218b

renanferrari commented on WordPress/src/main/java/org/wordpress/android/util/UriWrapper.kt:
```diff
@@ -1,15 +1,27 @@
     override fun toString() = uri.toString()
     fun getQueryParameter(key: String): String? = uri.getQueryParameter(key)
+    fun copy(path: String): UriWrapper {
+        val newUri = Builder()
```
I was almost sure there was a method to easily copy `Uri`s, so I just looked it up and this is the one: `Uri::buildUpon`. Have you considered using that? It seems like it would be safer.

planarvoid replied:
good catch! The previous approach was something used before, I've just copy&pasted it but I didn't know a method `buildUpon` exists. Changing it in in 2a83158eed

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java:
```diff
@@ -1171,6 +1171,14 @@ public static void showMainActivityAndSiteCreationActivity(Activity activity) {
+        intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
+        intent.putExtra(WPMainActivity.ARG_WP_COM_SIGN_UP, true);
+        activity.startActivity(intent);
+    }
```
I'm not sure if we need to add this. Calling `ActivityLauncher::showSignInForResult` with `isWpComOnly = true` should be enough in this case. Let me know what you think.

planarvoid replied:
I think you're right. I see that the signup flow triggers the WPMain activity itself, it doesn't just rely on the result. Why I haven't used it before is that the login activity is triggered by `startActivityForResult` and I expected that this is started from the Main activity and that the result is necessary (as it is for all the other deeplink flows). I've changed it in a1d4d99b4b

renanferrari reviewed (commented):
Again, thank you for the changes!

I've tested the app with the scenarios you provided and also the ones that were listed in #9957 and everything worked as expected.

Code is also looking pretty good. I just left a few comments below.

planarvoid committed:
Rename OpenInEditor to OpenEditorForPost

planarvoid committed:
Rename functions to explain they handle tracking URLs that don't have to come from emails

planarvoid committed:
Use buildUpon function instead of builder

planarvoid committed:
Use previous method showSignInForResult to trigger signin flow from deeplink

planarvoid committed:
Remove unused imports

renanferrari reviewed (approved):
Thanks for the changes again!

LGTM 👍
